# Introduction to `Pandas`

The `pandas` library is extremely useful for collecting, organizing, sorting, filtering, manipulating, and displaying large datasets.

The base object in the `pandas` library is the `DataFrame()`, which holds both the information and nearly all the necessary functions to manipulate it.  The library also allows the use of `display` instead of `print`, which makes the DataFrames look a little nicer and easier to read - and opens up additional formatting options later on!

First, let's begin by creating a simple DataFrame based on a dictionary of information.

In [3]:
import pandas as pd

my_dictionary = {"Name":["Bob","Carol","David","Elsie"],
                 "Age" :[48,22,99,55],
                 "Status":["Employed","Employed","Retired","Employed"],
                 "Salary":[65345,43214,0,78954]}

df = pd.DataFrame(my_dictionary)


In [4]:
display(df)

Name  Age    Status  Salary
0    Bob   48  Employed   65345
1  Carol   22  Employed   43214
2  David   99   Retired       0
3  Elsie   55  Employed   78954

First off, note the index numbers at the start of each row.  The index of each row is maintained until such time as they are either overwritten by a function or explicitly ignored during a function call.  This can be helpful when dealing with larger data sets or with unsorted data that needs to be sorted according to a specific set of criteria.

In a DataFrame object, we can treat column names like keys in a dictionary.

In [5]:
df["Name"]

0      Bob
1    Carol
2    David
3    Elsie
Name: Name, dtype: object

In [6]:
display(df["Name"])

0      Bob
1    Carol
2    David
3    Elsie
Name: Name, dtype: object

Notice that in both of the above cells, the information from a single column is displayed as an unformatted table.  However, each column is also treated like an object, and we can use some of the parameters inside the column objects to further extract information.  The `.values` parameter (note the LACK of parentheses there) returns a simple array or list of the values in that column, in order of row index.

In [7]:
df["Name"].values

array(['Bob', 'Carol', 'David', 'Elsie'], dtype=object)

We can pass a *list of names* instead of just a single name, and in that situation we will get a smaller DataFrame object back for display.  Pay careful attention here.  In the cell below, we have two sets of square brackets.  The outer brackets are the DataFrame's column selection, and the inner brackets signify a list containing the elements `"Name"` and `"Age"`.

In [8]:
display(df[["Name","Age"]])

Name  Age
0    Bob   48
1  Carol   22
2  David   99
3  Elsie   55

To see this a little more separated, we can divide the work up a bit.

In [9]:
columns_to_view = ["Name","Age"]
display( df[ columns_to_view ] )

Name  Age
0    Bob   48
1  Carol   22
2  David   99
3  Elsie   55

What if we wanted to add new data to our existing dataframe?  Perhaps a "Years to Retirement Age" column?  There are multiple ways of doing this.  First, we can provide a simple list of values in index order.  This means we need to already know both the values and the actual order.

In [10]:
years_to_retirement = [17,43,0,10]
df["Years to Retirement"] = years_to_retirement
display(df)

Name  Age    Status  Salary  Years to Retirement
0    Bob   48  Employed   65345                   17
1  Carol   22  Employed   43214                   43
2  David   99   Retired       0                    0
3  Elsie   55  Employed   78954                   10

Another option is to calculate the value of a column based on data in other columns.  Let's add a column that estimates how much total salary each employee will earn over the course of their remaining years to retirement.

*As an aside, datasets like these might be really boring, but they're also the simplest way to really illustrate the fundamentals before diving into the really powerful tools.*

We can assign a DataFrame column's value with a mathematical operation.  These can be simply or incredibly complex, so long as each row has all the necessary information to receive a value for the added column.

In [11]:
df["Estimated Payout"] = df["Salary"] * df["Years to Retirement"]
display(df)

Name  Age    Status  Salary  Years to Retirement  Estimated Payout
0    Bob   48  Employed   65345                   17           1110865
1  Carol   22  Employed   43214                   43           1858202
2  David   99   Retired       0                    0                 0
3  Elsie   55  Employed   78954                   10            789540

Way to go, Carol!

What if we wanted a more complicated formula, such as something that accounted for a 3% raise annually?

In [12]:
df["Final Salary"] = df["Salary"] * 1.03**df["Years to Retirement"]
display(df)

Name  Age    Status  Salary  Years to Retirement  Estimated Payout  \
0    Bob   48  Employed   65345                   17           1110865   
1  Carol   22  Employed   43214                   43           1858202   
2  David   99   Retired       0                    0                 0   
3  Elsie   55  Employed   78954                   10            789540   

    Final Salary  
0  108005.328531  
1  154037.027713  
2       0.000000  
3  106107.573815

*This brings us to a non-programming-related tip:
When job-hunting, arguing for a higher salary up front will ensure that every raise afterwards is based on a higher value, compounding your gains and making you more likely to continue being paid at the level appropriate to your skill.*

There are other ways the `pandas` library can load data into a DataFrame, such as from a file.

For this example, we'll be using a hydrogen bonding report generated by `cpptraj` on a molecular dynamics trajectory of the FusionRed fluorescent protein.  This data is shamelessly taken from former WalkerLab member Ashlyn Murphy's work (which was published already).

In the next cell, I will intentionally be using the same variable `df` for our DataFrame object.  It will overwrite the information previously held in there from earlier cells.  I'm doing this solely because most `pandas`-related documentation and examples use `df` by convention.

I'll be using the `.read_csv()` function, and because our file is not actually comma-separated, but whitespace-delimited, I'll be including the `delim_whitespace=True` keyword argument.

In [13]:
hbond_file = "../Datafiles/wt_hbonds.dat"
df = pd.read_csv(hbond_file,delim_whitespace=True)
display(df)

#Acceptor        DonorH        Donor  Frames    Frac  AvgDist  \
0     GLU_141@OE1   HIP_193@HE2  HIP_193@NE2   19364  0.9682   2.7489   
1     GLU_211@OE2   HIP_193@HD1  HIP_193@ND1   19059  0.9529   2.6854   
2       VAL_207@O      ILE_41@H     ILE_41@N   18994  0.9497   2.8163   
3       CH6_61@O2   ARG_88@HH22   ARG_88@NH2   18667  0.9334   2.8046   
4        MET_13@O      CYS_24@H     CYS_24@N   17959  0.8980   2.8178   
...           ...           ...          ...     ...     ...      ...   
1045    THR_104@O      PHE_84@H     PHE_84@N       1  0.0001   2.9911   
1046  ASP_147@OD2     GLY_148@H    GLY_148@N       1  0.0001   2.9925   
1047  THR_101@OG1  ARG_118@HH22  ARG_118@NH2       1  0.0001   2.9961   
1048    PHE_122@N   ASN_18@HD22   ASN_18@ND2       1  0.0001   2.9969   
1049    LEU_170@O  ASN_169@HD22  ASN_169@ND2       1  0.0001   2.9996   

        AvgAng  
0     158.8936  
1     161.3499  
2     161.9690  
3     163.3182  
4     158.4745  
...        ...  
1045  137.4207  
1046  141.1611  
1047  150.8644  
1048  143.6300  
1049  137.6103  

[1050 rows x 7 columns]

The original file already includes column names (indicated by the `#` at the start of the line), and `pandas` interprets those correctly, assigning each column appropriately.  You may also notice that the `display()` function shows five lines, then some dots, and then the last five lines.  At the bottom, we have a dimension report showing how many rows and columns are in our complete DataFrame.  This is default behavior for `pandas` when working with particularly large datasets, however it can be modified if necessary.  For now, we'll leave it as is.  Rest assured the information is not lost, even if it's hidden.

Now, this DataFrame has over a thousand rows.  What if we only wanted to look at hydrogen bond interactions that exist more than 85% (*chosen so we don't have 50 lines to show*) of the time?  We can simply slice our dataframe according to that condition.

In [14]:
df[ df["Frac"]>0.85 ]

#Acceptor       DonorH        Donor  Frames    Frac  AvgDist    AvgAng
0   GLU_141@OE1  HIP_193@HE2  HIP_193@NE2   19364  0.9682   2.7489  158.8936
1   GLU_211@OE2  HIP_193@HD1  HIP_193@ND1   19059  0.9529   2.6854  161.3499
2     VAL_207@O     ILE_41@H     ILE_41@N   18994  0.9497   2.8163  161.9690
3     CH6_61@O2  ARG_88@HH22   ARG_88@NH2   18667  0.9334   2.8046  163.3182
4      MET_13@O     CYS_24@H     CYS_24@N   17959  0.8980   2.8178  158.4745
5     LYS_159@O    LEU_134@H    LEU_134@N   17819  0.8910   2.8104  156.9981
6      LEU_11@O    SER_26@HG    SER_26@OG   17729  0.8864   2.7300  159.2946
7      LYS_67@O    ALA_215@H    ALA_215@N   17671  0.8835   2.8408  163.5637
8     LEU_150@O    TYR_174@H    TYR_174@N   17534  0.8767   2.8369  158.2550
9     VAL_191@O    GLU_141@H    GLU_141@N   17378  0.8689   2.8429  162.4089
10    LEU_158@O    LEU_166@H    LEU_166@N   17359  0.8679   2.8361  160.9188
11     GLU_27@O     ARG_40@H     ARG_40@N   17181  0.8590   2.8456  162.5429

We can also layer multiple conditions together.  Let's say we wanted to know which of these interactions ALSO had an average angle of more than 160.  We can combine conditional statements.  Note that in this process, we have to wrap each condition in parentheses.

In [15]:
df[ (df["Frac"]>0.85) & (df["AvgAng"]>160) ]

#Acceptor       DonorH        Donor  Frames    Frac  AvgDist    AvgAng
1   GLU_211@OE2  HIP_193@HD1  HIP_193@ND1   19059  0.9529   2.6854  161.3499
2     VAL_207@O     ILE_41@H     ILE_41@N   18994  0.9497   2.8163  161.9690
3     CH6_61@O2  ARG_88@HH22   ARG_88@NH2   18667  0.9334   2.8046  163.3182
7      LYS_67@O    ALA_215@H    ALA_215@N   17671  0.8835   2.8408  163.5637
9     VAL_191@O    GLU_141@H    GLU_141@N   17378  0.8689   2.8429  162.4089
10    LEU_158@O    LEU_166@H    LEU_166@N   17359  0.8679   2.8361  160.9188
11     GLU_27@O     ARG_40@H     ARG_40@N   17181  0.8590   2.8456  162.5429

Of the original 1,050 hydrogen bond interactions reported in the file, only seven of them are present more than 85% of the time and have an average angle greater than 160.  You can see how the `pandas` library can be useful for filtering large amounts of information down to more manageable chunks.

Let's consider a comparison between two systems.  For this section, we'll delete the `df` variable to clear up some memory, and then we'll use two new variable names for our wildtype `wt` and mutant `mut` data sets.

In [16]:
del df
wt  = pd.read_csv("../Datafiles/wt_hbonds.dat",delim_whitespace=True)
mut = pd.read_csv("../Datafiles/mut_hbonds.dat",delim_whitespace=True)

We can display each DataFrame individually, but what if we wanted to compare the values in a single table?  After all, when doing hydrogen bonding analyses, often the important information is what changed from one system to the next.  We can do this using the `merge()` function.  By default, the DataFrame being used to call `.merge()` is the "Left" DataFrame, and the first argument in the function is the "Right" DataFrame.  This is important because there are a number of keywords you can include which will determine how the system will respond in cases of ambiguity.

In [17]:
display(wt)

#Acceptor        DonorH        Donor  Frames    Frac  AvgDist  \
0     GLU_141@OE1   HIP_193@HE2  HIP_193@NE2   19364  0.9682   2.7489   
1     GLU_211@OE2   HIP_193@HD1  HIP_193@ND1   19059  0.9529   2.6854   
2       VAL_207@O      ILE_41@H     ILE_41@N   18994  0.9497   2.8163   
3       CH6_61@O2   ARG_88@HH22   ARG_88@NH2   18667  0.9334   2.8046   
4        MET_13@O      CYS_24@H     CYS_24@N   17959  0.8980   2.8178   
...           ...           ...          ...     ...     ...      ...   
1045    THR_104@O      PHE_84@H     PHE_84@N       1  0.0001   2.9911   
1046  ASP_147@OD2     GLY_148@H    GLY_148@N       1  0.0001   2.9925   
1047  THR_101@OG1  ARG_118@HH22  ARG_118@NH2       1  0.0001   2.9961   
1048    PHE_122@N   ASN_18@HD22   ASN_18@ND2       1  0.0001   2.9969   
1049    LEU_170@O  ASN_169@HD22  ASN_169@ND2       1  0.0001   2.9996   

        AvgAng  
0     158.8936  
1     161.3499  
2     161.9690  
3     163.3182  
4     158.4745  
...        ...  
1045  137.4207  
1046  141.1611  
1047  150.8644  
1048  143.6300  
1049  137.6103  

[1050 rows x 7 columns]

In [18]:
display(mut)

#Acceptor       DonorH        Donor  Frames    Frac  AvgDist    AvgAng
0     GLU_141@OE1  HIP_193@HE2  HIP_193@NE2   19411  0.9706   2.7340  157.7498
1     ASP_203@OD1  THR_205@HG1  THR_205@OG1   19017  0.9508   2.6528  166.8403
2       VAL_207@O     ILE_41@H     ILE_41@N   18758  0.9379   2.8198  161.9892
3       CH6_61@O2  ARG_88@HH22   ARG_88@NH2   18612  0.9306   2.8046  162.7157
4       LEU_150@O    TYR_174@H    TYR_174@N   18559  0.9280   2.8151  161.8743
...           ...          ...          ...     ...     ...      ...       ...
1068  GLU_151@OE2    GLY_152@H    GLY_152@N       1  0.0001   2.9881  146.3172
1069    THR_104@O     PHE_84@H     PHE_84@N       1  0.0001   2.9898  153.4486
1070  HID_210@NE2    VAL_212@H    VAL_212@N       1  0.0001   2.9904  140.9139
1071   GLU_27@OE2   LYS_10@HZ3    LYS_10@NZ       1  0.0001   2.9952  169.6782
1072     GLY_35@O  GLN_37@HE22   GLN_37@NE2       1  0.0001   2.9981  141.4157

[1073 rows x 7 columns]

The wildtype system shows 1,050 HBIs, while the mutant shows 1,073.  However, if we look at the first row of each, we can see they both start with the same Acceptor, DonorH, and Donor values.  Since we're doing a direct comparison between exact atom pairs, we'll use these three columns as the merge criteria.  All rows that have the same values in each of these three columns will be combined.  Let's see how that looks.

In [19]:
wt.merge(mut, on=["#Acceptor","DonorH","Donor"],how="outer",suffixes=["_wt","_mut"])

#Acceptor        DonorH        Donor  Frames_wt  Frac_wt  AvgDist_wt  \
0     GLU_141@OE1   HIP_193@HE2  HIP_193@NE2    19364.0   0.9682      2.7489   
1     GLU_211@OE2   HIP_193@HD1  HIP_193@ND1    19059.0   0.9529      2.6854   
2       VAL_207@O      ILE_41@H     ILE_41@N    18994.0   0.9497      2.8163   
3       CH6_61@O2   ARG_88@HH22   ARG_88@NH2    18667.0   0.9334      2.8046   
4        MET_13@O      CYS_24@H     CYS_24@N    17959.0   0.8980      2.8178   
...           ...           ...          ...        ...      ...         ...   
1268     THR_25@O    LYS_42@HZ1    LYS_42@NZ        NaN      NaN         NaN   
1269   ARG_216@NE  ASN_190@HD21  ASN_190@ND2        NaN      NaN         NaN   
1270   GLN_78@NE2   LYS_178@HZ1   LYS_178@NZ        NaN      NaN         NaN   
1271  GLU_151@OE2     GLY_152@H    GLY_152@N        NaN      NaN         NaN   
1272  HID_210@NE2     VAL_212@H    VAL_212@N        NaN      NaN         NaN   

      AvgAng_wt  Frames_mut  Frac_mut  AvgDist_mut  AvgAng_mut  
0      158.8936     19411.0    0.9706       2.7340    157.7498  
1      161.3499         NaN       NaN          NaN         NaN  
2      161.9690     18758.0    0.9379       2.8198    161.9892  
3      163.3182     18612.0    0.9306       2.8046    162.7157  
4      158.4745     17913.0    0.8956       2.8218    159.0163  
...         ...         ...       ...          ...         ...  
1268        NaN         1.0    0.0001       2.9835    151.9126  
1269        NaN         1.0    0.0001       2.9855    144.1207  
1270        NaN         1.0    0.0001       2.9860    139.3766  
1271        NaN         1.0    0.0001       2.9881    146.3172  
1272        NaN         1.0    0.0001       2.9904    140.9139  

[1273 rows x 11 columns]

The first thing we may notice is that there are several rows with `NaN` in them. This is short for "Not a Number", and is put in wherever there is no data to merge.  It is very difficult to perform any kind of arithmetic with `NaN` values, but fortunately for us, `pandas` has a solution in place.  We will simply attach another function to the end of the merge call, `.fillna()`.

In [20]:
combined = wt.merge(mut, on=["#Acceptor","DonorH","Donor"],how="outer",suffixes=["_wt","_mut"]).fillna(0)
display(combined)

#Acceptor        DonorH        Donor  Frames_wt  Frac_wt  AvgDist_wt  \
0     GLU_141@OE1   HIP_193@HE2  HIP_193@NE2    19364.0   0.9682      2.7489   
1     GLU_211@OE2   HIP_193@HD1  HIP_193@ND1    19059.0   0.9529      2.6854   
2       VAL_207@O      ILE_41@H     ILE_41@N    18994.0   0.9497      2.8163   
3       CH6_61@O2   ARG_88@HH22   ARG_88@NH2    18667.0   0.9334      2.8046   
4        MET_13@O      CYS_24@H     CYS_24@N    17959.0   0.8980      2.8178   
...           ...           ...          ...        ...      ...         ...   
1268     THR_25@O    LYS_42@HZ1    LYS_42@NZ        0.0   0.0000      0.0000   
1269   ARG_216@NE  ASN_190@HD21  ASN_190@ND2        0.0   0.0000      0.0000   
1270   GLN_78@NE2   LYS_178@HZ1   LYS_178@NZ        0.0   0.0000      0.0000   
1271  GLU_151@OE2     GLY_152@H    GLY_152@N        0.0   0.0000      0.0000   
1272  HID_210@NE2     VAL_212@H    VAL_212@N        0.0   0.0000      0.0000   

      AvgAng_wt  Frames_mut  Frac_mut  AvgDist_mut  AvgAng_mut  
0      158.8936     19411.0    0.9706       2.7340    157.7498  
1      161.3499         0.0    0.0000       0.0000      0.0000  
2      161.9690     18758.0    0.9379       2.8198    161.9892  
3      163.3182     18612.0    0.9306       2.8046    162.7157  
4      158.4745     17913.0    0.8956       2.8218    159.0163  
...         ...         ...       ...          ...         ...  
1268     0.0000         1.0    0.0001       2.9835    151.9126  
1269     0.0000         1.0    0.0001       2.9855    144.1207  
1270     0.0000         1.0    0.0001       2.9860    139.3766  
1271     0.0000         1.0    0.0001       2.9881    146.3172  
1272     0.0000         1.0    0.0001       2.9904    140.9139  

[1273 rows x 11 columns]

Now we can see that these rows have 0.0000 in them, which means we can use these values for further math.  However, this DataFrame is starting to look a little crowded.  We can use the `.drop()` function to drop entire columns from our table.  Keep in mind that unless you capture the result of the function into a variable (a new one or the same one) the changes will be **displayed only**.  For now, let's drop the distances, angles, and frames.  We'll still have the `"Frac"` columns for both, and since the total duration of each original trajectory was the same (20,000 frames), we know that the fractions represent the same number of frames.  In addition, this comparison is easier if the systems had slightly different numbers of frames, because the overall percentage is what we're interested in anyway.

In [21]:
combined.drop(["AvgDist_mut","AvgDist_wt","AvgAng_wt","AvgAng_mut","Frames_wt","Frames_mut"],axis=1)

#Acceptor        DonorH        Donor  Frac_wt  Frac_mut
0     GLU_141@OE1   HIP_193@HE2  HIP_193@NE2   0.9682    0.9706
1     GLU_211@OE2   HIP_193@HD1  HIP_193@ND1   0.9529    0.0000
2       VAL_207@O      ILE_41@H     ILE_41@N   0.9497    0.9379
3       CH6_61@O2   ARG_88@HH22   ARG_88@NH2   0.9334    0.9306
4        MET_13@O      CYS_24@H     CYS_24@N   0.8980    0.8956
...           ...           ...          ...      ...       ...
1268     THR_25@O    LYS_42@HZ1    LYS_42@NZ   0.0000    0.0001
1269   ARG_216@NE  ASN_190@HD21  ASN_190@ND2   0.0000    0.0001
1270   GLN_78@NE2   LYS_178@HZ1   LYS_178@NZ   0.0000    0.0001
1271  GLU_151@OE2     GLY_152@H    GLY_152@N   0.0000    0.0001
1272  HID_210@NE2     VAL_212@H    VAL_212@N   0.0000    0.0001

[1273 rows x 5 columns]

Now we'll save the changes above into yet another new DataFrame (again to make it easier and to reduce the number of things we have to do at each step.  Then, we'll add a new column which subtracts the wildtype fraction from the mutant fraction.  This will show us what changes when we mutate the protein.  Negative values indicate the mutant lost hydrogen bonding interactions compared to the wildtype, positive values indicate a gain of interaction.

In [22]:
cleaned_df = combined.drop(["AvgDist_mut","AvgDist_wt","AvgAng_wt","AvgAng_mut","Frames_wt","Frames_mut"],axis=1)
cleaned_df["Difference"] = cleaned_df["Frac_mut"] - cleaned_df["Frac_wt"]
display(cleaned_df)

#Acceptor        DonorH        Donor  Frac_wt  Frac_mut  Difference
0     GLU_141@OE1   HIP_193@HE2  HIP_193@NE2   0.9682    0.9706      0.0024
1     GLU_211@OE2   HIP_193@HD1  HIP_193@ND1   0.9529    0.0000     -0.9529
2       VAL_207@O      ILE_41@H     ILE_41@N   0.9497    0.9379     -0.0118
3       CH6_61@O2   ARG_88@HH22   ARG_88@NH2   0.9334    0.9306     -0.0028
4        MET_13@O      CYS_24@H     CYS_24@N   0.8980    0.8956     -0.0024
...           ...           ...          ...      ...       ...         ...
1268     THR_25@O    LYS_42@HZ1    LYS_42@NZ   0.0000    0.0001      0.0001
1269   ARG_216@NE  ASN_190@HD21  ASN_190@ND2   0.0000    0.0001      0.0001
1270   GLN_78@NE2   LYS_178@HZ1   LYS_178@NZ   0.0000    0.0001      0.0001
1271  GLU_151@OE2     GLY_152@H    GLY_152@N   0.0000    0.0001      0.0001
1272  HID_210@NE2     VAL_212@H    VAL_212@N   0.0000    0.0001      0.0001

[1273 rows x 6 columns]

This is certainly easier to read without the extra columns - and that data is still completely available in the previous DataFrame objects or simply by reloading the original datafile.

Since our goal here is to compare the different hydrogen bonding interactions between two systems, we might want to filter our DataFrame by the `Difference` column.  Let's look at values that differ by more than 30% of the simulation time.  This means any interactions that increase **or** decrease by 30%.  Numerically, that means we want values in the `Difference` column that are below -0.3 or above 0.3, with nothing in between.  We can use the `abs()` function in base python to get the absolute value of the column and filter that way.

In [23]:
filtered_df = cleaned_df[abs(cleaned_df["Difference"]) > 0.3]
display(filtered_df)

#Acceptor        DonorH        Donor  Frac_wt  Frac_mut  Difference
1     GLU_211@OE2   HIP_193@HD1  HIP_193@ND1   0.9529    0.0000     -0.9529
18    GLU_211@OE1     ARG_194@H    ARG_194@N   0.8231    0.0000     -0.8231
43      TYR_174@O  ARG_175@HH11  ARG_175@NH1   0.7300    0.0000     -0.7300
56      CYS_154@O     LEU_170@H    LEU_170@N   0.6835    0.0000     -0.6835
60      ALA_153@O     THR_142@H    THR_142@N   0.6742    0.0000     -0.6742
64       GLN_37@O     GLU_211@H    GLU_211@N   0.6613    0.0000     -0.6613
76      THR_142@O     ALA_153@H    ALA_153@N   0.6238    0.0000     -0.6238
80      LEU_170@O     CYS_154@H    CYS_154@N   0.6122    0.0000     -0.6122
95      PHE_122@O   ASN_18@HD22   ASN_18@ND2   0.5513    0.0000     -0.5513
96    GLU_211@OE1     VAL_212@H    VAL_212@N   0.5406    0.0000     -0.5406
103     SER_79@OG    SER_176@HG   SER_176@OG   0.5227    0.0533     -0.4694
105     CH6_61@OH    SER_139@HG   SER_139@OG   0.5056    0.0118     -0.4938
117     GLU_211@O    THR_36@HG1   THR_36@OG1   0.4717    0.0000     -0.4717
118   GLU_151@OE1   THR_173@HG1  THR_173@OG1   0.4711    0.0434     -0.4277
122    GLN_78@OE1     LYS_184@H    LYS_184@N   0.4399    0.0328     -0.4071
133   ASN_190@OD1   THR_142@HG1  THR_142@OG1   0.4061    0.0408     -0.3653
142   GLU_151@OE2   THR_173@HG1  THR_173@OG1   0.3825    0.0559     -0.3266
150     GLU_211@O      GLN_37@H     GLN_37@N   0.3623    0.0000     -0.3623
151   GLU_151@OE2  ARG_175@HH22  ARG_175@NH2   0.3535    0.0000     -0.3535
157   GLU_151@OE1  ARG_175@HH22  ARG_175@NH2   0.3380    0.0000     -0.3380
163   GLU_151@OE2  ARG_175@HH12  ARG_175@NH1   0.3167    0.0000     -0.3167
165   GLU_151@OE1  ARG_175@HH12  ARG_175@NH1   0.3079    0.0000     -0.3079
194     ASP_155@O    SER_139@HG   SER_139@OG   0.2099    0.9134      0.7035
269      ILE_55@O    THR_58@HG1   THR_58@OG1   0.0877    0.4400      0.3523
475     VAL_212@O     ARG_194@H    ARG_194@N   0.0089    0.3778      0.3689
1050     GLN_37@O     GLH_211@H    GLH_211@N   0.0000    0.7742      0.7742
1051  GLH_211@OE1   HIP_193@HD1  HIP_193@ND1   0.0000    0.6565      0.6565
1052    ILE_154@O     LEU_170@H    LEU_170@N   0.0000    0.6404      0.6404
1053    LEU_170@O     ILE_154@H    ILE_154@N   0.0000    0.5970      0.5970
1054    PHE_122@O   ASN_18@HD21   ASN_18@ND2   0.0000    0.5786      0.5786
1055    ASN_153@O     THR_142@H    THR_142@N   0.0000    0.5154      0.5154
1056    GLH_211@O   GLH_211@HE2  GLH_211@OE2   0.0000    0.4975      0.4975
1057    THR_142@O     ASN_153@H    ASN_153@N   0.0000    0.4233      0.4233
1058  ASN_153@OD1  ASN_169@HD21  ASN_169@ND2   0.0000    0.4158      0.4158

As luck would have it (or just the nature of math), the rows are almost perfectly sorted in increasing order by the value of the difference column.  However, there are a few rows that are out of place in this ordering, so let's enforce that arrangement.  We can use the `.sort_values()` function with the keyword argument `ascending=True` to put the smallest value at the top and the largest value at the bottom.  We'll include the column we want to sort by as well.

In [24]:
filtered_df.sort_values(by="Difference",ascending=True)

#Acceptor        DonorH        Donor  Frac_wt  Frac_mut  Difference
1     GLU_211@OE2   HIP_193@HD1  HIP_193@ND1   0.9529    0.0000     -0.9529
18    GLU_211@OE1     ARG_194@H    ARG_194@N   0.8231    0.0000     -0.8231
43      TYR_174@O  ARG_175@HH11  ARG_175@NH1   0.7300    0.0000     -0.7300
56      CYS_154@O     LEU_170@H    LEU_170@N   0.6835    0.0000     -0.6835
60      ALA_153@O     THR_142@H    THR_142@N   0.6742    0.0000     -0.6742
64       GLN_37@O     GLU_211@H    GLU_211@N   0.6613    0.0000     -0.6613
76      THR_142@O     ALA_153@H    ALA_153@N   0.6238    0.0000     -0.6238
80      LEU_170@O     CYS_154@H    CYS_154@N   0.6122    0.0000     -0.6122
95      PHE_122@O   ASN_18@HD22   ASN_18@ND2   0.5513    0.0000     -0.5513
96    GLU_211@OE1     VAL_212@H    VAL_212@N   0.5406    0.0000     -0.5406
105     CH6_61@OH    SER_139@HG   SER_139@OG   0.5056    0.0118     -0.4938
117     GLU_211@O    THR_36@HG1   THR_36@OG1   0.4717    0.0000     -0.4717
103     SER_79@OG    SER_176@HG   SER_176@OG   0.5227    0.0533     -0.4694
118   GLU_151@OE1   THR_173@HG1  THR_173@OG1   0.4711    0.0434     -0.4277
122    GLN_78@OE1     LYS_184@H    LYS_184@N   0.4399    0.0328     -0.4071
133   ASN_190@OD1   THR_142@HG1  THR_142@OG1   0.4061    0.0408     -0.3653
150     GLU_211@O      GLN_37@H     GLN_37@N   0.3623    0.0000     -0.3623
151   GLU_151@OE2  ARG_175@HH22  ARG_175@NH2   0.3535    0.0000     -0.3535
157   GLU_151@OE1  ARG_175@HH22  ARG_175@NH2   0.3380    0.0000     -0.3380
142   GLU_151@OE2   THR_173@HG1  THR_173@OG1   0.3825    0.0559     -0.3266
163   GLU_151@OE2  ARG_175@HH12  ARG_175@NH1   0.3167    0.0000     -0.3167
165   GLU_151@OE1  ARG_175@HH12  ARG_175@NH1   0.3079    0.0000     -0.3079
269      ILE_55@O    THR_58@HG1   THR_58@OG1   0.0877    0.4400      0.3523
475     VAL_212@O     ARG_194@H    ARG_194@N   0.0089    0.3778      0.3689
1058  ASN_153@OD1  ASN_169@HD21  ASN_169@ND2   0.0000    0.4158      0.4158
1057    THR_142@O     ASN_153@H    ASN_153@N   0.0000    0.4233      0.4233
1056    GLH_211@O   GLH_211@HE2  GLH_211@OE2   0.0000    0.4975      0.4975
1055    ASN_153@O     THR_142@H    THR_142@N   0.0000    0.5154      0.5154
1054    PHE_122@O   ASN_18@HD21   ASN_18@ND2   0.0000    0.5786      0.5786
1053    LEU_170@O     ILE_154@H    ILE_154@N   0.0000    0.5970      0.5970
1052    ILE_154@O     LEU_170@H    LEU_170@N   0.0000    0.6404      0.6404
1051  GLH_211@OE1   HIP_193@HD1  HIP_193@ND1   0.0000    0.6565      0.6565
194     ASP_155@O    SER_139@HG   SER_139@OG   0.2099    0.9134      0.7035
1050     GLN_37@O     GLH_211@H    GLH_211@N   0.0000    0.7742      0.7742

What if we wanted to sort by two columns, perhaps the `#Acceptor` and `Donor` columns?  Simple!  Instead of passing a single value to the `by` and `ascending` keywords, we'll pass lists.  Keep in mind that the order of the lists matters.  The first item in the list will be the primary column for sorting, and then any rows with the exact same value in that first column will then be sub-sorted by the second item in the list, and so forth.  The order also matters with the `ascending` keyword, as this ordering will match the list of columns.

In [25]:
filtered_df.sort_values(by=["#Acceptor","Donor"],ascending=[True,True])

#Acceptor        DonorH        Donor  Frac_wt  Frac_mut  Difference
60      ALA_153@O     THR_142@H    THR_142@N   0.6742    0.0000     -0.6742
1055    ASN_153@O     THR_142@H    THR_142@N   0.0000    0.5154      0.5154
1058  ASN_153@OD1  ASN_169@HD21  ASN_169@ND2   0.0000    0.4158      0.4158
133   ASN_190@OD1   THR_142@HG1  THR_142@OG1   0.4061    0.0408     -0.3653
194     ASP_155@O    SER_139@HG   SER_139@OG   0.2099    0.9134      0.7035
105     CH6_61@OH    SER_139@HG   SER_139@OG   0.5056    0.0118     -0.4938
56      CYS_154@O     LEU_170@H    LEU_170@N   0.6835    0.0000     -0.6835
1056    GLH_211@O   GLH_211@HE2  GLH_211@OE2   0.0000    0.4975      0.4975
1051  GLH_211@OE1   HIP_193@HD1  HIP_193@ND1   0.0000    0.6565      0.6565
1050     GLN_37@O     GLH_211@H    GLH_211@N   0.0000    0.7742      0.7742
64       GLN_37@O     GLU_211@H    GLU_211@N   0.6613    0.0000     -0.6613
122    GLN_78@OE1     LYS_184@H    LYS_184@N   0.4399    0.0328     -0.4071
165   GLU_151@OE1  ARG_175@HH12  ARG_175@NH1   0.3079    0.0000     -0.3079
157   GLU_151@OE1  ARG_175@HH22  ARG_175@NH2   0.3380    0.0000     -0.3380
118   GLU_151@OE1   THR_173@HG1  THR_173@OG1   0.4711    0.0434     -0.4277
163   GLU_151@OE2  ARG_175@HH12  ARG_175@NH1   0.3167    0.0000     -0.3167
151   GLU_151@OE2  ARG_175@HH22  ARG_175@NH2   0.3535    0.0000     -0.3535
142   GLU_151@OE2   THR_173@HG1  THR_173@OG1   0.3825    0.0559     -0.3266
150     GLU_211@O      GLN_37@H     GLN_37@N   0.3623    0.0000     -0.3623
117     GLU_211@O    THR_36@HG1   THR_36@OG1   0.4717    0.0000     -0.4717
18    GLU_211@OE1     ARG_194@H    ARG_194@N   0.8231    0.0000     -0.8231
96    GLU_211@OE1     VAL_212@H    VAL_212@N   0.5406    0.0000     -0.5406
1     GLU_211@OE2   HIP_193@HD1  HIP_193@ND1   0.9529    0.0000     -0.9529
1052    ILE_154@O     LEU_170@H    LEU_170@N   0.0000    0.6404      0.6404
269      ILE_55@O    THR_58@HG1   THR_58@OG1   0.0877    0.4400      0.3523
80      LEU_170@O     CYS_154@H    CYS_154@N   0.6122    0.0000     -0.6122
1053    LEU_170@O     ILE_154@H    ILE_154@N   0.0000    0.5970      0.5970
95      PHE_122@O   ASN_18@HD22   ASN_18@ND2   0.5513    0.0000     -0.5513
1054    PHE_122@O   ASN_18@HD21   ASN_18@ND2   0.0000    0.5786      0.5786
103     SER_79@OG    SER_176@HG   SER_176@OG   0.5227    0.0533     -0.4694
76      THR_142@O     ALA_153@H    ALA_153@N   0.6238    0.0000     -0.6238
1057    THR_142@O     ASN_153@H    ASN_153@N   0.0000    0.4233      0.4233
43      TYR_174@O  ARG_175@HH11  ARG_175@NH1   0.7300    0.0000     -0.7300
475     VAL_212@O     ARG_194@H    ARG_194@N   0.0089    0.3778      0.3689

You might notice something about the new ordering.  We specified the `#Acceptor` column for our first sorting element.  From a biochemistry standpoint, we might be thinking about amino acids in sequence order rather than alphabetical order.  But `pandas` doesn't know that, it only knows the data we've given it.

Fortunately, we can use some python magic to sort based on a **portion** of the column we're working with.  We'll use the base python function `sorted()` to do this.  But first, we'll need to work out how to get the specific portion we want for each row, and we'll want to ensure we're sorting it **numerically**, not alphabetically (in code, the digits 0-9 can be either numbers OR 'letters', so it's important to be aware of how you're treating them).

First, the `sorted()` function itself looks like this:

```python
my_list = [7,4,3,5,2,6,1]
sorted_list = sorted(my_list)
```

This is fine for a list of single values.  However, if we have a list of lists, we'll need to be clear which element of each list we're sorting on.

```python
my_list = [[1,9],[2,3],[3,8],[4,7],[6,5],[8,0],[0,4]]
sorted_list = sorted(my_list,key=lambda x: x[0])
```
We've introduced the `key` argument to our function, which tells python what to do with each element in the list by way of sorting things.  In this example, we've told it to sort by the first element (index 0) in each part of our larger list.  The resulting sorted list will be in ascending order by the first element, with the respective second element moved accordingly.

Our pandas DataFrame column has a string of letters, numbers, and symbols, so we'll need to specify what portion of that string we want to use.  Let's look at an example of how we can identify the particular portion of the string we want.  We can chain multiple string `.split()` functions together along with list indices, to get the final result.  Let's look at it piece-wise first, then connect it all together.



In [26]:
my_string = "ALA_153@O"
substring = my_string.split("_")[1] # index 1, or the second element in the list returned from the split("_") function.
print(substring,type(substring))
subsubstring = substring.split("@")[0] # index 0, or the first element of the list returned from the split("@") function.
print(subsubstring,type(subsubstring))
### Keep in mind, subsubstring is still a string variable, not a number.  To treat it numerically, we need to cast it inside `int()`
my_res_id = int(subsubstring)
print(my_res_id,type(my_res_id))

153@O <class 'str'>
153 <class 'str'>
153 <class 'int'>


Now let's combine all of those individual `split()` and `int()` functions and casts together.

In [27]:
my_string = "ALA_153@O"
my_res_id = int(my_string.split("_")[1].split("@")[0])
print(my_res_id,type(my_res_id))

153 <class 'int'>


Effectively, this just takes all the intermediate variables and replaces them with the values assigned to them in the first place.  Now, let's apply this same process to our DataFrame.  In the cell below, we're creating a new DataFrame using the sorted values from our `filtered_df`, using a key that specifies the first element of our row (index 0 is the first column, which is our `#Acceptor` column), and then using the `split()` and `int()` pieces.  Then we include the columns in the new DataFrame as well.

In [28]:
new_df = pd.DataFrame(
    sorted(filtered_df.values, key=lambda x: int(x[0].split('@')[0].split("_")[1])), 
    columns=filtered_df.columns
)
display(new_df)

#Acceptor        DonorH        Donor  Frac_wt  Frac_mut  Difference
0      GLN_37@O     GLU_211@H    GLU_211@N   0.6613    0.0000     -0.6613
1      GLN_37@O     GLH_211@H    GLH_211@N   0.0000    0.7742      0.7742
2      ILE_55@O    THR_58@HG1   THR_58@OG1   0.0877    0.4400      0.3523
3     CH6_61@OH    SER_139@HG   SER_139@OG   0.5056    0.0118     -0.4938
4    GLN_78@OE1     LYS_184@H    LYS_184@N   0.4399    0.0328     -0.4071
5     SER_79@OG    SER_176@HG   SER_176@OG   0.5227    0.0533     -0.4694
6     PHE_122@O   ASN_18@HD22   ASN_18@ND2   0.5513    0.0000     -0.5513
7     PHE_122@O   ASN_18@HD21   ASN_18@ND2   0.0000    0.5786      0.5786
8     THR_142@O     ALA_153@H    ALA_153@N   0.6238    0.0000     -0.6238
9     THR_142@O     ASN_153@H    ASN_153@N   0.0000    0.4233      0.4233
10  GLU_151@OE1   THR_173@HG1  THR_173@OG1   0.4711    0.0434     -0.4277
11  GLU_151@OE2   THR_173@HG1  THR_173@OG1   0.3825    0.0559     -0.3266
12  GLU_151@OE2  ARG_175@HH22  ARG_175@NH2   0.3535    0.0000     -0.3535
13  GLU_151@OE1  ARG_175@HH22  ARG_175@NH2   0.3380    0.0000     -0.3380
14  GLU_151@OE2  ARG_175@HH12  ARG_175@NH1   0.3167    0.0000     -0.3167
15  GLU_151@OE1  ARG_175@HH12  ARG_175@NH1   0.3079    0.0000     -0.3079
16    ALA_153@O     THR_142@H    THR_142@N   0.6742    0.0000     -0.6742
17    ASN_153@O     THR_142@H    THR_142@N   0.0000    0.5154      0.5154
18  ASN_153@OD1  ASN_169@HD21  ASN_169@ND2   0.0000    0.4158      0.4158
19    CYS_154@O     LEU_170@H    LEU_170@N   0.6835    0.0000     -0.6835
20    ILE_154@O     LEU_170@H    LEU_170@N   0.0000    0.6404      0.6404
21    ASP_155@O    SER_139@HG   SER_139@OG   0.2099    0.9134      0.7035
22    LEU_170@O     CYS_154@H    CYS_154@N   0.6122    0.0000     -0.6122
23    LEU_170@O     ILE_154@H    ILE_154@N   0.0000    0.5970      0.5970
24    TYR_174@O  ARG_175@HH11  ARG_175@NH1   0.7300    0.0000     -0.7300
25  ASN_190@OD1   THR_142@HG1  THR_142@OG1   0.4061    0.0408     -0.3653
26  GLU_211@OE2   HIP_193@HD1  HIP_193@ND1   0.9529    0.0000     -0.9529
27  GLU_211@OE1     ARG_194@H    ARG_194@N   0.8231    0.0000     -0.8231
28  GLU_211@OE1     VAL_212@H    VAL_212@N   0.5406    0.0000     -0.5406
29    GLU_211@O    THR_36@HG1   THR_36@OG1   0.4717    0.0000     -0.4717
30    GLU_211@O      GLN_37@H     GLN_37@N   0.3623    0.0000     -0.3623
31  GLH_211@OE1   HIP_193@HD1  HIP_193@ND1   0.0000    0.6565      0.6565
32    GLH_211@O   GLH_211@HE2  GLH_211@OE2   0.0000    0.4975      0.4975
33    VAL_212@O     ARG_194@H    ARG_194@N   0.0089    0.3778      0.3689

Now we can see that even though the values are not in alphabetically ascending order, they are in numerical order by residue number from the `#Acceptor` column.  Now let's do it with multiple columns as we tried before.

In [29]:
new_df = pd.DataFrame(
    sorted(filtered_df.values, key=lambda x: [int(x[0].split('@')[0].split("_")[1]),int(x[2].split('@')[0].split("_")[1])]), 
    columns=filtered_df.columns
)
display(new_df)

#Acceptor        DonorH        Donor  Frac_wt  Frac_mut  Difference
0      GLN_37@O     GLU_211@H    GLU_211@N   0.6613    0.0000     -0.6613
1      GLN_37@O     GLH_211@H    GLH_211@N   0.0000    0.7742      0.7742
2      ILE_55@O    THR_58@HG1   THR_58@OG1   0.0877    0.4400      0.3523
3     CH6_61@OH    SER_139@HG   SER_139@OG   0.5056    0.0118     -0.4938
4    GLN_78@OE1     LYS_184@H    LYS_184@N   0.4399    0.0328     -0.4071
5     SER_79@OG    SER_176@HG   SER_176@OG   0.5227    0.0533     -0.4694
6     PHE_122@O   ASN_18@HD22   ASN_18@ND2   0.5513    0.0000     -0.5513
7     PHE_122@O   ASN_18@HD21   ASN_18@ND2   0.0000    0.5786      0.5786
8     THR_142@O     ALA_153@H    ALA_153@N   0.6238    0.0000     -0.6238
9     THR_142@O     ASN_153@H    ASN_153@N   0.0000    0.4233      0.4233
10  GLU_151@OE1   THR_173@HG1  THR_173@OG1   0.4711    0.0434     -0.4277
11  GLU_151@OE2   THR_173@HG1  THR_173@OG1   0.3825    0.0559     -0.3266
12  GLU_151@OE2  ARG_175@HH22  ARG_175@NH2   0.3535    0.0000     -0.3535
13  GLU_151@OE1  ARG_175@HH22  ARG_175@NH2   0.3380    0.0000     -0.3380
14  GLU_151@OE2  ARG_175@HH12  ARG_175@NH1   0.3167    0.0000     -0.3167
15  GLU_151@OE1  ARG_175@HH12  ARG_175@NH1   0.3079    0.0000     -0.3079
16    ALA_153@O     THR_142@H    THR_142@N   0.6742    0.0000     -0.6742
17    ASN_153@O     THR_142@H    THR_142@N   0.0000    0.5154      0.5154
18  ASN_153@OD1  ASN_169@HD21  ASN_169@ND2   0.0000    0.4158      0.4158
19    CYS_154@O     LEU_170@H    LEU_170@N   0.6835    0.0000     -0.6835
20    ILE_154@O     LEU_170@H    LEU_170@N   0.0000    0.6404      0.6404
21    ASP_155@O    SER_139@HG   SER_139@OG   0.2099    0.9134      0.7035
22    LEU_170@O     CYS_154@H    CYS_154@N   0.6122    0.0000     -0.6122
23    LEU_170@O     ILE_154@H    ILE_154@N   0.0000    0.5970      0.5970
24    TYR_174@O  ARG_175@HH11  ARG_175@NH1   0.7300    0.0000     -0.7300
25  ASN_190@OD1   THR_142@HG1  THR_142@OG1   0.4061    0.0408     -0.3653
26    GLU_211@O    THR_36@HG1   THR_36@OG1   0.4717    0.0000     -0.4717
27    GLU_211@O      GLN_37@H     GLN_37@N   0.3623    0.0000     -0.3623
28  GLU_211@OE2   HIP_193@HD1  HIP_193@ND1   0.9529    0.0000     -0.9529
29  GLH_211@OE1   HIP_193@HD1  HIP_193@ND1   0.0000    0.6565      0.6565
30  GLU_211@OE1     ARG_194@H    ARG_194@N   0.8231    0.0000     -0.8231
31    GLH_211@O   GLH_211@HE2  GLH_211@OE2   0.0000    0.4975      0.4975
32  GLU_211@OE1     VAL_212@H    VAL_212@N   0.5406    0.0000     -0.5406
33    VAL_212@O     ARG_194@H    ARG_194@N   0.0089    0.3778      0.3689

Now that we have the data sorted by both the `#Acceptor` and `Donor` columns, we can more easily examine our data.  Remember that we took two complete hydrogen bonding analysis outputs from `cpptraj`, and through a series of sorting, merging, and filtering steps, we got down to just 33 lines that meet our criteria of "Interactions that differ by more than 30%".  Now we can start applying our chemical knowledge to what we see here to decide what's really important.  

Rows 16 and 17 above are both between residue 153 and 142, though they show different amino acid codes for 153.  This is because that position was mutated as part of the project.  If we look at the `Frac` columns for those two rows, we see that each row has a 0 value in either the `_wt` or `_mut`.  Simply put, these are zeroes because those exact interactions couldn't have existed in both systems - the residues had different names.  However, if we're wanting to see what changes between systems, we actually see that the fraction goes from 0.67 to 0.52, which is actually less than our 30% threshold.  It's possible we might not really care about this smaller change.

At rows 19 and 20, we see a similar effect with the changing of the amino acid.  The actual numerical change for the hydrogen bonds between these two positions is only about 0.04, putting that change well below our threshold of 30%.

Now look at rows 28 and 29.  GLU/GLH are the same amino acid, except that one is protonated and the other is not.  Our knowledge of biochemistry also tells us that the OE1 and OE2 atoms in a glutamate are identical to each other, which means that the difference between systems might not actually be as pronounced as the `Difference` column would suggest.  In fact, the actual change is only about 30% when you consider the identity of the atoms involved, not 95% or 65% from the two rows individually.

Now, let's look at formatting of our tables.  Let's say we wanted to highlight cells based on their value.  We've already sorted and filtered our data, but now we want to identify which values have changed the most.  Let's highlight interactions that decrease by 60% in red, and those that increase by 60% in green.

First, we have to define a function that will return the proper color style information for a given cell.  Then, we use the `.style.applymap()` function and pass the name of our highlighting function as well as the specific columns we're applying the function to.  The result should have some cells highlighted in red or green.

In [30]:
def highlight_cells(val):
    if float(val) > 0.6:
        color = "green"
    elif float(val) < -0.6:
        color = "red"
    else:
        color = ""
    return 'background-color: {}'.format(color)

new_df.style.applymap(highlight_cells,subset=["Difference"])

We can also do something similar to highlight entire rows based on the value of a specific column.  This will use the `.style.apply()` function (note the difference here from the previous one!)  We'll pass an `axis=1` keyword argument to the function to let `pandas` know we're working at a row level, not a single-cell level.

In [31]:
def highlight_row(row):
    val = row["Difference"]
    if float(val) > 0.6:
        color = "green"
    elif float(val) < -0.6:
        color = "red"
    else:
        color = "grey"
    return ['background-color: {}'.format(color) for r in row]  
    # Notice that we're returning a list of individual style items, because the row is like a list and we want to highlight the whole thing.

new_df.style.apply(highlight_row,axis=1)


One of the last things we'll cover is saving our dataframes as a file.  Depending on your needs, a simple CSV file might be enough, or you might want to save your data as an Excel spreadsheet with multiple sheets inside.  Fortunately, `pandas` has functionality for a number of output styles.  First is the CSV.  By default, the separator is a comma (`,`), however we can modify this to another character if we prefer.  Below, I've used `\t`, the `tab` character, in our `.to_csv()` function call.  Also, we don't want the row indices in our CSV file, as that can cause errors when reading the file back in at a later time.  So we'll include the `index=False` keyword as well.

In [32]:
new_df.to_csv("../Datafiles/pandas.csv",sep="\t",index=False)

The CSV file we've created doesn't have any formatting like we demonstrated above.  To save cell formatting, we'll need a more complicated file type, like an Excel spreadsheet.
Also, the `new_df` variable doesn't have the formatting saved to it from above.  Instead, what we did in the cells above merely applied a display style without saving it.  Let's save a copy of that DataFrame with the formatting included.

In [33]:
styled_df = new_df.style.apply(highlight_row,axis=1)
styled_df.to_excel("../Datafiles/pandas.xlsx",index=False)

If we open the `.xlsx` file we just created, we'll see that it has our data and formatting all in place.  This can be very useful when working with larger datasets, especially.

What if we had multiple dataframes that we wanted to save into a single excel file across multiple sheets?  If we tried the method above, we'd just keep overwriting each previous step with the next one.  So instead, we'll use an `ExcelWriter` object available in `pandas`.  With this, we can add things to the excel file in stages, then save and close it when we're finished.  The cell below uses an intermediate variable `writer` to keep the `ExcelWriter` information in, and once all the indented lines have finished below it, the `ExcelWriter` saves the file and closes it, allowing the user to then open it with other programs or send it off to a PI to satisfy their hunger for **more data**.

In [34]:
with pd.ExcelWriter('../Datafiles/pandas_multi_sheet.xlsx') as writer:  
    new_df.to_excel(writer, sheet_name='Unformatted DataFrame')
    styled_df.to_excel(writer, sheet_name='Fancy Formatted DataFrame')

The next output format is perhaps a bit less common, but it is still worth knowing about, especially if you find yourself working with $\LaTeX$ for papers and other documents (such as the textbook associated with this workshop).  In simple form, the `.to_latex()` function can be called without any arguments, and the result is just a string that you can manipulate, print to the console, or write to a text file.  The cell below shows the DataFrame contained in a `tabular` $\LaTeX$ environment.  Note that it does a good job of including escapes (`\`) in places with special characters that would otherwise confuse a $\LaTeX$ compiler, such as the `#` in `#Acceptor` or the `_` in every row's first three columns.

In [38]:
latex_data = new_df.to_latex(index=False)
new_df.to_latex("../Datafiles/pandas.tex",index=False)
print(latex_data)

\begin{tabular}{lllrrr}
\toprule
  \#Acceptor &       DonorH &       Donor &  Frac\_wt &  Frac\_mut &  Difference \\
\midrule
   GLN\_37@O &    GLU\_211@H &   GLU\_211@N &   0.6613 &    0.0000 &     -0.6613 \\
   GLN\_37@O &    GLH\_211@H &   GLH\_211@N &   0.0000 &    0.7742 &      0.7742 \\
   ILE\_55@O &   THR\_58@HG1 &  THR\_58@OG1 &   0.0877 &    0.4400 &      0.3523 \\
  CH6\_61@OH &   SER\_139@HG &  SER\_139@OG &   0.5056 &    0.0118 &     -0.4938 \\
 GLN\_78@OE1 &    LYS\_184@H &   LYS\_184@N &   0.4399 &    0.0328 &     -0.4071 \\
  SER\_79@OG &   SER\_176@HG &  SER\_176@OG &   0.5227 &    0.0533 &     -0.4694 \\
  PHE\_122@O &  ASN\_18@HD22 &  ASN\_18@ND2 &   0.5513 &    0.0000 &     -0.5513 \\
  PHE\_122@O &  ASN\_18@HD21 &  ASN\_18@ND2 &   0.0000 &    0.5786 &      0.5786 \\
  THR\_142@O &    ALA\_153@H &   ALA\_153@N &   0.6238 &    0.0000 &     -0.6238 \\
  THR\_142@O &    ASN\_153@H &   ASN\_153@N &   0.0000 &    0.4233 &      0.4233 \\
GLU\_151@OE1 &  THR\_173@HG1 & THR

As we've seen, `pandas` is a very powerful and useful library that can help you organize, sort, filter, manipulate, present, and save large datasets.

I'm not sure why it's named `pandas`, though.